primap2/pm2io/_data_reading.py

```diff
@@ -1490,64 +1523,105 @@ def harmonize_units(
+                        data_this_entity = data.loc[data[entity_col] == entity]
+                        units_this_entity = data_this_entity[unit_col].unique()
+
+                        for unit in units_this_entity:
+                            if unit != unit_to:
+                                # print(f"Working on unit {unit}")
+                                unit_pint = ureg[unit]
+                                # could add a try except block here to throw and log an
+                                # error or add error info in DF instead of crashing
+                                gwp_this_entity = basic_entities[basic_entity][entity]
+                                if gwp_this_entity:
+                                    with ureg.context(gwp_this_entity):
+                                        unit_pint = unit_pint.to(unit_to)
+                                else:
+                                    unit_pint = unit_pint.to(unit_to)
+                                # print(f"Pint unit is {unit_pint}")
+                                factor = unit_pint.magnitude
+                                # print(f"Converting with factor {factor} to unit
+                                # {unit_to}")
+                                mask = (data[entity_col] == entity) & (
+                                    data[unit_col] == unit
+                                )
+                                # print(data.loc[mask, data_cols])
+                                try:
+                                    data.loc[mask, data_cols] *= factor
+                                except TypeError:
+                                    # print(data.loc[mask, data_cols])
+                                    strs = find_str_values_in_data(data, data_cols)
+                                    logger.error(
+                                        f"The following string values are present and "
+                                        f"can not be converted during unit conversion: "
+                                        f"{strs}."
+                                    )
+                                    raise ValueError(
+                                        f"String values {strs} prevent unit conversion."
+                                    ) from None
+
+                                data.loc[mask, unit_col] = unit_to
+
+                        # if entity differs from basic entity and the units are not
+                        # compatible we had GWP conversion and have to adapt the entity
+                        if (entity != basic_entity) and not ureg(
+                            unit
+                        ).is_compatible_with(ureg[unit_to]):
+                            # entity was converted
+                            entity_mask = data[entity_col] == entity
+                            # print(f"Changing entity from {entity} to {basic_entity}")
+                            data.loc[entity_mask, entity_col] = basic_entity
```

Comment:
Uh, I haven't read all that. I guess it is fine? If it is critical, I'd need to take an hour or four to understand this absolute beast of a function again (and probably slice it into 3 or 4 functions in the process). If you are reasonably confident it at least can't lead to *wrong* results (suboptimal result units are fine, I mean wrong as in converts to one unit but tells in the result it was another unit), I'd skip that for now.